### DISCUSSION TITLE
Performance of clear commands in example

### DISCUSSION THREAD

**Discussion by Desperado17** [DiligentGraphics/DiligentSamples#105] [Q&A]
Greetings,

I tried to use the Diligent examples to make some performance comparisons between Vulkan and OpenGL. For this purpose, I used Tutorial01_HelloTriangle because it's simple:

https://github.com/DiligentGraphics/DiligentSamples/blob/master/Tutorials/Tutorial01_HelloTriangle/src/Tutorial01_HelloTriangle.cpp

I started the application with and without the "-m GL" parameter on a native Ubuntu Linux X86. Then I measured the timestamps for the Vulkan/OpenGL commands using the Performance Counter View tab of RenderDoc:

https://renderdoc.org/

I tried this on 3 different machines: Nvidia with proprietary driver, Intel with proprietary driver, Intel via Mesa. In all cases, the clear commands glClear vs. vkCmdClear* yielded significantly shorter timestamps with OpenGL than with Vulkan usually 30% shorter or less. All in all, the more pixels a command produced the worse Vulkan performed. It only regularly won on the indexed draw calls.

Anyone an idea why that is so?

Regards.

**Comment by TheMostDiligent** [comment]
There are multiple ways a buffer may be cleared in Vulkan:
- Clear outside of render pass
- Clear the attachment when render pass starts
- Clear the attachment within the render pass

The performance of these methods depend on the GPU architecture, driver, etc. There is only one command in GL and depending on the circumstances it may map to one of these methods.

For the 30% better time - is this the time of clear command only? Note that sometimes it may be difficult to reliably measure the duration of a single command.

> It only regularly won on the indexed draw calls.

Are you talking about the performance of indexed draw calls or the clear performance when combined with indexed draw call?

**Reply by Desperado17** [reply]
What I can observe is that

a) vkCmdClearAttachments performs worse than glClear

b) vkCmdDrawIndexed performs better than glDrawElements

On a different test of mine, a full screen quad was slower with vkCmdDraw than glDrawArrays.
It seems to me that the more pixels are involved the worse the Vulkan implementations perform.

**Reply by TheMostDiligent** [reply]
You can also try to clear with the render pass LOAD_OP_CLEAR operation.

> On a different test of mine, a full screen quad was slower with vkCmdDraw than glDrawArrays.

How did you measure it?

**Reply by Desperado17** [reply]
Render doc performance counter view shows times for commands. So does Nvidia Nsight when capturing a frame with frame profiler.

**Reply by TheMostDiligent** [reply]
Not sure why vkCmdDraw may be slower than glDrawArrays really. Maybe try some computationally extensive workload.

**Reply by Desperado17** [reply]
Do you know if it is common for vkCmdClearAttachments to be slower than glClear?

**Reply by TheMostDiligent** [reply]
That depends on the GPU, driver, implementation, etc.

glClear is likely mapped to one of the 3 clear methods in Vulkan. Clearing with LOAD_OP_CLEAR is likely the fastest method in Vulkan, would be interesting to compare glClear against it

**Reply by Desperado17** [reply]
Just to give you some numbers:

For Tutorial01_HelloTriangle the first vkCmdClearAttachments takes 44.512 musec on gpu while the corresponding glClearBufferfv takes 23.104 musec. The vkCmdClearAttachments for the depth/stencil takes 45.152 on Vulkan while the corresponding glClear ond Opengl takes 3.744 musec. On the other hand, the vkCmdDraw takes 11.68 musec on Vulkan and the corresponding glDrawArrays takes 18.624. The final vkCmdDrawIndexed takes 7.104 musec on Vulkan and glDrawElements takes 1.056 musec.

I get similar results on 3 different Ubuntu desktops with Nvidia native driver and/or Mesa.

**Reply by TheMostDiligent** [reply]
I don't know if such short durations are reliable. There may be some fixed overhead that may prevail in such cases.

**Reply by Desperado17** [reply]
In a larger application that I'm porting from OpenGL to Vulkan removing the most pixel expensive drawcalls led to an immense speedup. From three dozen drawcalls, removing two screen filling drawcalls raised the FPS from 400 to 500 in an instant. Do the examples have an FPS count option for the console or something? Title bar is too fast.

**Reply by Desperado17** [reply]
Nevertheless, I think the Vulkan mode reaches up to 3k something frames while OpenGL reaches up to 4k. Compiled in ReleaseWIthDbg.

**Reply by TheMostDiligent** [reply]
What GPU are you testing on?

**Reply by Desperado17** [reply]
Nvidia Quadro M1000M  with [510.73.05]( driver on Ubuntu 18.04.

**Reply by TheMostDiligent** [reply]
I don't have a definite answer, but I still think that when measuring such short draw calls, some side effects may prevail the timing.

**Reply by Desperado17** [reply]
Even cumulative over multiple frames? And a fullscreen clear should be expensive enough, shouldn't it be?

**Reply by Desperado17** [reply]
Do you get a higher framerate for Vulkan on your side and if yes which driver?

**Reply by TheMostDiligent** [reply]
There are certain more complicated scenarios that run better on Vulkan (such as Tutorials 6, 9, 10, 16). Other applications run on par with OpenGL.
I have NVidia RTX2080 and also older NVidia GTX970.

**Reply by Desperado17** [reply]
So, how are the framerates on these systems for Triangle01?

**Reply by TheMostDiligent** [reply]
On My RTX2080, Tutorial 1 runs at ~3.1 ms/frame in Vulkan mode and ~4.2 ms/frame in OpenGL mode (Windows). I don't have any explanation for the difference other than internal driver implementation.  I also don't think these are very representative results.

**Reply by Desperado17** [reply]
May I ask for the cmake command that you used to generate the windows engine plus examples? I think I have a colleague with windows hardware who could also try it.

**Reply by TheMostDiligent** [reply]
https://github.com/DiligentGraphics/DiligentEngine#win32

**Reply by Desperado17** [reply]
Unfortunately, the only colleague of mine that has a Windows boot can't build there. Would it be possible to obtain the binary from you? The Triangle01 windows plus the two libs?

**Reply by TheMostDiligent** [reply]
[Tutorial01.zip](https://github.com/DiligentGraphics/DiligentSamples/files/9444799/Tutorial01.zip)

**Reply by Desperado17** [reply]
Thanks!

**Reply by Desperado17** [reply]
Ok, had a colleague with Windows and Intel HD Graphics 530 test this. OpenGL: 700 fps, Vulkan 400 fps.

I wonder if there is a certain optimization new chips hagve old don't.

**Reply by Desperado17** [reply]
Windows 10 with Quadro RTX 4000. Vulkan wins over OpenGL with 0.32 ms per frame vs. 0.39 ms.

**Reply by Desperado17** [reply]
Btw some Linux systems limit the Vulkan frames to 60 fps. Can you imagine why?

**Reply by TheMostDiligent** [reply]
I definitely see that Vulkan works better or modern GPUs.

> Btw some Linux systems limit the Vulkan frames to 60 fps. Can you imagine why?

The driver may not support the required present mode, but I don't know why.

**Reply by Desperado17** [reply]
I tried on two Linux systems, one of which went to 60fps for triangle01, the other not. both present the

Diligent Endinge: Warning: VK_PRESENT_MODE_MAILBOX_KHR is not supported. Defaulting to VK_PRESENT_MODE_FIFO_KHR message.

So I have doubts this is the cause.

**Reply by TheMostDiligent** [reply]
I'm not sure then

**Reply by Desperado17** [reply]
Is there a simple way to make the examples default to VK_PRESENT_MODE_IMMEDIATE_KHR?

**Reply by TheMostDiligent** [reply]
If you disable VSync, it will try MAILBOX->IMMEDIATE->FIFO

**Reply by Desperado17** [reply]
Is there a simple way maybe without recompiling?

**Reply by TheMostDiligent** [reply]
![image](https://user-images.githubusercontent.com/61806567/187469236-88e983ca-9ac7-428f-ba34-588439668330.png)

**Reply by Desperado17** [reply]
Is there a command line parameter like "-vsync off"   ? He keeps choosing FIFO.

**Reply by Desperado17** [reply]
No matter if Vsync is checked or not, no more than 60 FPS on the Ubuntu 22 with High End NVidia.

**Reply by TheMostDiligent** [reply]
I don't know why it does not disable VSync. Maybe the required present mode is not supported by the driver.
